### PR TITLE
MBS-13757: Support unchecking recordings to update

### PR DESCRIPTION
**Pull request by derat**
# MBS-13757

Update the release editor so that individual tracks' checkboxes to update recording titles or artists can be unchecked after the "copy all" checkboxes have been checked.

# Problem

Updating most (but not all) of the recording titles/artists on a release (common when fixing compilations) currently requires a bunch of clicking. There's no way to uncheck individual tracks after clicking one of the "copy all" checkboxes, so you need to individually click each of the tracks whose recordings should be updated. It's easy to make mistakes here since each track may have two checkboxes with similarly-worded text (for updating the title vs. artist).

# Solution

Stop disabling the per-track checkboxes when a "copy all" checkbox is checked, and make the "copy all" checkboxes use computed values so they can be updated in response to the per-track checkboxes.

The "copy all" checkboxes are automatically checked if all of the tracks are checked and unchecked otherwise. This code feels a bit strange to me since tracks have hidden state even when their checkboxes are hidden because they already match their recordings, but that was already the case before this change.

# Testing

Manually tested using my dev server.

### CODE REVIEW

**Comment by derat** [comment]
@mwiencek and @reosarevok, mind taking a look?

@Aerozol, curious to get your thoughts too.

**Comment by Aerozol** [comment]
Yes please! I use a userscript to do this already:
"[MB: QoL: Select All Update Recordings](https://github.com/ROpdebee/mb-userscripts)" by ROpdebee

Which gives:
image

But to be honest, I also don't quite get the benefit in toggling them all and then locking them (e.g. the current default behavious), so it makes sense to me to just change the default behaviour. I've only used the userscript button, probably for years. Perhaps there is something I am missing, or a discussion that was had back in MB history re. why the checkboxes were decided to be locked - curious to see what bitmap and reo think.

**Comment by derat** [comment]
@Aerozol, thanks -- it sounds like that userscript was created for the same reason as this PR. :-)

Does the userscript's "Toggle all" button invert the state of each of the checkboxes? If so, do you ever use that button? I'm not sure of the use case for it (vs. a "Deselect all" button).

**Comment by derat** [comment]
@brainzbot, retest this please

**Comment by Aerozol** [comment]
Correct, toggle inverts the selection - I've never used it.

It's cool for a userscript, but unless there's a use case I haven't thought of I prefer the simpler UI of just one button, for official implementation.

**Comment by reosarevok** [comment]
> I also don't quite get the benefit in toggling them all and then locking them (e.g. the current default behaviour)

If I change some track titles, check the "change all" recording checkbox, then change some more track titles, the subsequent changes will also be automatically selected and applied to recordings. If the checkbox was a one-off toggle, I expect any further changes would need to be re-toggled separately.

**Comment by derat** [comment]
@reosarevok, did you get a chance to try this to see how it feels?

**Comment by reosarevok** [comment]
Ok, I finally tested it now - my only worry is whether it's confusing to have a case where I manually deselect one track, so the "change all" checkbox is no longer checked, but then I change another track and that still shows as checked - that might be unintuitive if the main checkbox is now unchecked. That said, I don't mind that it works like that, just wonder if there's a better way to indicate it than just unchecking the main checkbox. I could see keeping it checked but with a warning icon that says "some tracks were manually unchecked" but that might be too much effort for the gain.

**Comment by derat** [comment]
Thanks for taking a look! Yeah, the hidden checked state for not-yet-changed tracks seems like the tricky part to me too.

Maybe it would be too complicated/magic, but one option might be to also uncheck unchanged tracks along with the copy-all checkbox when the user manually unchecks a track. I'm uploading another commit with that behavior to make it easier to compare.

Another option might be adding a "partially checked" state (e.g. via partial opacity) to the copy-all checkboxes that gets used when a track has been manually unchecked, but I suspect that that might feel even weirder...

**Comment by derat** [comment]
I updated this to use an `allTracks` generator function and to make tracks subscribe to `copyTrackTitlesToRecordings` and `copyTrackArtistsToRecordings` (I'm not sure what the best practices are here).

**Comment by derat** [comment]
@mwiencek, would you mind taking a look at this too? Thanks!

File: root/static/scripts/release-editor/fields.js [hunk 1]
```diff
@@ -1231,6 +1239,28 @@ class Release extends mbEntity.Release {
...
+  // Returns a generator for iterating over all tracks.
+  allTracks() {
+    return (function* () {
+      for (const medium of this.mediums()) {
+        for (const track of medium.tracks()) {
+          yield track;
+        }
+      }
+    }).apply(this);
+  }
```
Comment (mwiencek) [inline root/static/scripts/release-editor/fields.js:1251]: Is there a reason having the star on `*allTracks()` wouldn't work? That should still return a fresh generator each time it's called IIUC.
```suggestion
  *allTracks() {
    for (const medium of this.mediums()) {
      for (const track of medium.tracks()) {
        yield track;
      }
    }
  }
```

**Review by mwiencek — APPROVED** [review]
Sorry for the delay! I tested this locally and it works perfectly.

File: root/static/scripts/release-editor/fields.js (hunk 1 above)
Comment (derat) [reply root/static/scripts/release-editor/fields.js:1251]: Thanks! I'm new to generator functions and had actually searched for e.g. `javascript method generator` to see if something like this was possible, but I apparently missed that the method syntax is actually documented near the bottom of <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*>. :-P

(eslint requires a space after the asterisk, which looks a bit weird, but oh well.)

**Pull request merged**